arthur/exts/grafana/ldap_team_sync.py

```diff
@@ -0,0 +1,169 @@
+            )
+            added_members += 1
+        return MissingMembers(count=len(missing_members), successfully_added=added_members)
+
+    async def _remove_extra_members(
+        self,
+        grafana_team_id: int,
+        ldap_team_members: set[str],
+        grafana_team_members: set[str],
+        all_grafana_users: list[dict],
+    ) -> int:
+        """
+        Removes Grafana users from a team if they are not present in the LDAP team.
+
+        Return how many were removed.
+        """
+        extra_members = grafana_team_members - ldap_team_members
+        removed_members = 0
+        for grafana_user in all_grafana_users:
+            if grafana_user["login"] not in extra_members:
+                continue
+            await grafana.remove_user_from_team(
+                grafana_user["userId"],
+                grafana_team_id,
+                self.bot.http_session,
+            )
+            removed_members += 1
+        return removed_members
+
+    async def _sync_teams(self, team: dict[str, str]) -> SyncFigures:
+        """
+        Ensure members in LDAP are present in Grafana teams.
+
+        Return the number of members missing from the Grafana team, and the number of members added.
+        """
+        if team["name"] not in GRAFANA_TO_LDAP_NAME_MAPPING:
+            return SyncFigures(added=MissingMembers(0, 0), removed=0)
+
+        ldap_team_members = {
+            member.uid
+            for member in await ldap.get_group_members(GRAFANA_TO_LDAP_NAME_MAPPING[team["name"]])
+        }
+        grafana_team_members = {
+            member["login"]
+            for member in await grafana.list_team_members(team["id"], self.bot.http_session)
+            if member.get("auth_module") == "ldap"
+        }
+
+        all_grafana_users = await grafana.get_all_users(self.bot.http_session)
+        added_members = await self._add_missing_members(
+            team["id"],
+            ldap_team_members,
+            grafana_team_members,
+            all_grafana_users,
+        )
+        removed_members = await self._remove_extra_members(
+            team["id"],
+            ldap_team_members,
+            grafana_team_members,
+            all_grafana_users,
+        )
+
+        return SyncFigures(added=added_members, removed=removed_members)
+
+    @tasks.loop(hours=12)
+    async def sync_ldap_grafana_teams(self, channel: discord.TextChannel | None = None) -> None:
+        """Update Grafana team membership to match LDAP team membership."""
+        grafana_teams = await grafana.list_teams(self.bot.http_session)
+        embed = discord.Embed(
+            title="Sync Stats",
+            colour=discord.Colour.blue(),
+        )
+        for team in grafana_teams:
+            logger.debug(f"Processing {team["name"]}")
+            try:
+                figures = await self._sync_teams(team)
+            except aiohttp.ClientResponseError as e:
+                logger.error(e)
```

Comment:
This is not the right way to log the error. It turns the exception into a string (likely with some incomplete information, like the key in a `KeyError`) and hides the traceback.
Instead, do `logger.error("We screwed up!", exc_info=e)`

Reply:
Or is it different in Loguru?

Reply:
That should work in loguru too.

`logger.exception("We screwed up!")` is an alias for that too, where it automatically adds exc_info based on context. (Python's logging module has the same)

Reply:
Actually yeah, I missed the fact that `loguru` is used. But yeah the issue is similar
```ipython
In [3]: logger.error(KeyError('debugging adventure'))
2024-07-26 17:22:20.344 | ERROR    | __main__:<module>:1 - 'debugging adventure'
```
So the right replacement (according to the [docs](https://loguru.readthedocs.io/en/stable/resources/migration.html#replacing-exc-info-argument)) would be
```ipython
In [6]: logger.opt(exception=KeyError('debugging adventure')).error('Describe what happened')
2024-07-26 17:24:51.429 | ERROR    | __main__:<module>:1 - Describe what happened
KeyError: 'debugging adventure'
```